Development practices/folder_structure.md

```diff
@@ -0,0 +1,286 @@
+    ├── get_driving_distance.rb
+    ├── get_geo_keywords.rb
+    ├── get_location_attributes.rb
+    ├── get_route.rb
+    └── leave_reachable.rb
+```
+
+</details>
+
+At some point this folder became a dumpall folder without rhyme or reason.
+There is also an issue of boundries: there are none. Each class is used from anywhere in the application and that can
+cause headaches.
+
+# Desired structure
+
+Here is a folder structure we should strive for
+
+```
+app
+├── controllers  # Rails 
+├── frontend     # webpacker
+├── domains      # custom
+├── models       # Rails
+├── views        # Rails
+└── web          # custom
+```
+
+Lets first talk about Rails specific folders (`controllers`, `views` and `models`).
+
+`views` folder needs to exist because it is hardcoded in the Rails itself (https://github.com/rails/rails/blob/main/railties/lib/rails/engine.rb#L610)
+
+`controllers` is there because of the way rails finds controller names (it appends `"Controller"` to the resource name) (https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/http/request.rb#L88)
+
+`models` is a folder where we put all our ActiveRecord models. It is technically not necessary but we will talk about it in [Further Improvements](#further-improvements) chapter.
+
+
+# `domains` vs `web` folder
+`web` is for classes used in controller actions: forms, policies, serializers, queries and similar.
+`domains` is for business logic.
+
+## `web` folder
+`web` folder structure should follow your routes. Having these routes:
+
+```ruby
+namespace: :dasboard do
+  resources :projects
+end
+namespace :api do
+  namespace :v1 do
+    resources :projects
+  end
+end
+```
+
+it would have this folder structure:
+
+```
+web
+├── dashboard
+│   ├── projects
+│   │   ├── presenter.rb
+│   │   ├── decorator.rb
+│   │   ├── form.rb
+│   │   └── policy.rb
+├── api
+│   ├── v1
+│   │   ├── v1
+│   │   │   ├── projects
+│   │   │   │   ├── policy.rb
+│   │   │   │   ├── query.rb
+│   │   │   │   └── serialzier.rb
+```
+
+## `domains` folder
+`domains` folder should be structured into domains. 
+
+The main idea is to create cohesive groups of domain objects that are involved in the same business use-case(s). The same way we try to build classes that are loosely coupled but highly cohesive we should build our domains as well.
+Here is an example of domains folder:
+
+```
+domains
+├── exports
+│   ├── formatters
+│   ├── csvs
+│   └── spreadsheets
+├── external
+│   └── zapier.rb
+├── projects
+│   ├── counter.rb
+│   ├── mailer.rb
+│   └── presister.rb
+├── utils
+│   ├── cache.rb
+│   └── slug.rb
+├── application_decorator.rb
+├── application_mailer.rb
+├── application_policy.rb
+├── application_presenter.rb
+├── application_query.rb
+├── application_serializer.rb
+```
+
+## Further improvements
+* Move controllers to web folder as well
+* Move models to domains and web folders (build DTOs)
+* Move views to web folder
+* Move frontend to web folder
```

Comment:
Basically, the end goal is to have just the `web` and `domains` folders? I understand the idea behind moving these folders to the `web` folder, but it seems like then it will become a dumpall folder.

Reply:
I would be awesome if we could have a `app` and a `web` folder in the root, but dont know how and if we can make Rails do that :D

Reply:
I wouldn't say it's a dumpall folder. We have it on RGO where we have a bunch of models/domains and it works really nicely for us - it's really organized and you know where to look for something.

Reply:
@nikajukic could you elaborate what exactly you have on RGO? I'm looking on staging branch and see controllers and models in their normal rails places.

I'm not a fan of these "improvements", especially for models and controllers. 
From looking at these two folders I can see how "big" the app is, and know exactly where what is.
 
I don't see the benefit for most of our apps. If we go down that road, we will probably have multiple iterations and improvements over time, resulting in different structures on different projects - more time to onboard on each project.

Reply:
This is something I would love to experiment with, not something we should do :) 
And we will never know if we don't try :D

Reply:
Sorry, I meant that in web folder we have everything organized into resource folders that basically correspond to models we have (exept controllers and actual models).

E.g. report_requirement and authority are our models

web
-- report_requirement
---- serializer.rb
---- query.rb
---- filter/some_filter.rb
-- authority
---- serializer.rb
---- policy.rb



We have models and controllers set up the Rails way, we didn't want to hack our way around it to be able to move them to web/domains folder.